### PR TITLE
[AArch64][GISel] Add fp128 and i128 sitofp/uitofp handling

### PR DESCRIPTION
**Pull request by Him188**
Legalize sitofp/uitofp involving fp128/i128 types into a libcall. 
Vector with i128/fp128 types are scalarized.

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/itofp.ll [hunk 1]
```diff
@@ -740,395 +741,571 @@ entry:
...
-; CHECK-NEXT:    fmov d0, d8
-; CHECK-NEXT:    ldr x30, [sp, #16] // 8-byte Folded Reload
-; CHECK-NEXT:    fmov d1, d9
-; CHECK-NEXT:    ldp d9, d8, [sp], #64 // 16-byte Folded Reload
-; CHECK-NEXT:    ret
-entry:
-  %c = uitofp <3 x i128> %a to <3 x double>
-  ret <3 x double> %c
-}
-
```
Comment (dc03-work) [inline llvm/test/CodeGen/AArch64/itofp.ll:1131]: Why were these two tests deleted?

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 2]
```diff
@@ -710,34 +710,45 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
...
       .widenScalarIf(
           [=](const LegalityQuery &Query) {
-            return Query.Types[0].getScalarSizeInBits() >
-                   Query.Types[1].getScalarSizeInBits();
+            return Query.Types[1].getScalarSizeInBits() <= 64 &&
+                   Query.Types[0].getScalarSizeInBits() <
+                       Query.Types[1].getScalarSizeInBits();
           },
-          LegalizeMutations::changeElementSizeTo(1, 0))
+          LegalizeMutations::changeElementSizeTo(0, 1))
```
Comment (dc03-work) [inline llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:746]: I don't think it matters in practice (or at all really), but why were the `changeElementSizeTo` calls swapped?

**Review by dc03-work — COMMENTED** [review]
Seems like codegen matches SDAG now (and is better for the tail-call cases). Nice.

File: llvm/test/CodeGen/AArch64/itofp.ll (hunk 1 above)
Comment (Him188) [reply llvm/test/CodeGen/AArch64/itofp.ll:1131]: Hmm... That was a mistake when resolving rebase conflicts... Will add them back.

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 3]
```diff
@@ -1136,16 +1136,13 @@ LegalizerHelper::libcall(MachineInstr &MI, LostDebugLocObserver &LocObserver) {
   }
   case TargetOpcode::G_SITOFP:
   case TargetOpcode::G_UITOFP: {
-    // FIXME: Support other types
     unsigned FromSize = MRI.getType(MI.getOperand(1).getReg()).getSizeInBits();
-    unsigned ToSize = MRI.getType(MI.getOperand(0).getReg()).getSizeInBits();
-    if ((FromSize != 32 && FromSize != 64) || (ToSize != 32 && ToSize != 64))
+    Type *ToTy =
+        getFloatTypeForLLT(Ctx, MRI.getType(MI.getOperand(0).getReg()));
+    if ((FromSize != 32 && FromSize != 64 && FromSize != 128) || !ToTy)
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:1142]: Some targets might want to extend from 16 bit floats. It seems to be forbidden.

Comment (Him188) [reply llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:1142]: This is a similar check like the one in the G_FPTOUI above. I think it's better to forbid 16-bit types as they are not tested. If other targets want 16-bit floats, they can easily relax this check.

**Comment by Him188** [comment]
Ping @davemgreen and @aemerson ?

**Review by davemgreen — APPROVED** [review]
This looks OK, and looks very similar to the fp->int conversions. I'm not sure if there is a way to improve the i128->f16 convert without needing to introduce other libcalls.

LGTM

**Pull request merged**